### DISCUSSION TITLE
PCL2无法正确地自动识别部分原生启动器安装的愚人节版本 #3360

### DISCUSSION THREAD

**Discussion by allMagicNB** [Meloong-Git/PCL#3361] [讨论]
#3360

**Comment by allMagicNB** [comment]
我上传下Json（

**Reply by allMagicNB** [reply]
[20w14infinite.json](https://github.com/Hex-Dragon/PCL2/files/14322248/20w14infinite.json)
[3D Shareware v1.34.json](https://github.com/Hex-Dragon/PCL2/files/14322250/3D.Shareware.v1.34.json)
[22w13oneblockatatime.json](https://github.com/Hex-Dragon/PCL2/files/14322251/22w13oneblockatatime.json)
[23w13a_or_b.json](https://github.com/Hex-Dragon/PCL2/files/14322255/23w13a_or_b.json)
[1.RV-Pre1.json](https://github.com/Hex-Dragon/PCL2/files/14322256/1.RV-Pre1.json)

**Comment by mc-124** [comment]
好 找到办法了
原版端JSON的releaseTime好像不会变，遍历version_manifest.json[versions]，获取键releaseTime的值和原版端的releaseTime对比即可
建议读version_manifest.json而不是version_manifest_v2.json，v2比v1大，下载要花更多时间

**Reply by hejiehao** [reply]
这似乎需要联网

**Comment by mc-124** [comment]
> 这似乎需要联网

PCL2不会缓存是吧

**Reply by hejiehao** [reply]
万一有人每次启动 PCL2 都没网呢？

**Comment by mc-124** [comment]
> 万一有人每次启动 PCL2 都没网呢？

那直接摆烂显示id得了
谁会玩mc了还总没网

**Comment by mc-124** [comment]
> 好 找到办法了 原版端JSON的releaseTime好像不会变，遍历version_manifest.json[versions]，获取键releaseTime的值和原版端的releaseTime对比即可 建议读version_manifest.json而不是version_manifest_v2.json，v2比v1大，下载要花更多时间

发现部分mc的json被转换过utf+，需要先将releaseTime改为+00:00

**Comment by mc-124** [comment]
口误，UTC+（`七根木棍做的东西`有毛病缺js改不了）

**Comment by wuliaodexiaoluo** [comment]
> 那直接摆烂显示id得了 谁会玩mc了还总没网

你可能需要考虑到的情况：
MCBBS 挂了/维护（现在就是）/关闭
BMCLAPI 挂了（之前有）/维护/关闭
Mojang API 挂了/维护/关闭/被墙（部分地区现在就是）

**Reply by Deep-Dark-Forest** [reply]
Mojang API会挂？我简直不敢相信

**Reply by hejiehao** [reply]
但是你会挂

**Comment by mc-124** [comment]
> > 那直接摆烂显示id得了 谁会玩mc了还总没网
> 
> 你可能需要考虑到的情况： MCBBS 挂了/维护（现在就是）/关闭 BMCLAPI 挂了（之前有）/维护/关闭 Mojang API 挂了/维护/关闭/被墙（部分地区现在就是）

![image](https://github.com/Hex-Dragon/PCL2/assets/138775126/efb411e9-31f7-47cd-9d89-64a39266d755)

**Comment by mc-124** [comment]
这是大概的代码样子，龙猫改改就应该可以套上去了（
```vb
Imports System
Imports Newtonsoft.Json
Imports System.Globalization

''' <summary>
''' 根据提供的Minecraft JSON数据和版本列表JSON数据，找到匹配的Minecraft版本信息
''' </summary>
''' <param name="mcj">Minecraft的JSON数据对象，包含需要查找的版本信息</param>
''' <param name="vj">包含所有Minecraft版本信息的JSON数据对象</param>
''' <returns>如果找到匹配的版本，则返回该版本的JSON字符串；如果未找到，则抛出异常</returns>
Public Function FindMinecraftVersion(mcj As Object, vj As Object) As String
    Dim mcv As Object = Nothing '找到的mc版本 为Nothing表示未找到
    Dim releaseTime As String = DateTime.Parse(mcj("releaseTime").ToString(), Nothing, DateTimeStyles.AdjustToUniversal).ToString() '获取到的JSOn内的releaseTime

    For Each version As Object In vj("versions")
        If releaseTime = DateTime.Parse(version("releaseTime").ToString(), Nothing, DateTimeStyles.AdjustToUniversal).ToString() Then
            mcv = version
            Exit For
        End If
    Next

    If mcv IsNot Nothing Then
        Return JsonConvert.SerializeObject(mcv)
    Else
        Throw New Exception("无法确定Minecraft版本")
    End If
End Function
```

**Comment by wuliaodexiaoluo** [comment]
> Mojang API会挂？我简直不敢相信

是服务器就肯定会挂，给API压个几亿的访问或者压个1T的流量会触发CDN流量清洗

**Reply by wuliaodexiaoluo** [reply]
或者环境雪崩（软件卡死出错）

**Comment by Zhichii** [comment]
@allMagicNB 你不是来过 #3294 吗，你看看那个方法行不行
![image](https://github.com/Hex-Dragon/PCL2/assets/69286920/158a2638-e169-4a0f-bb5e-875c11e9cea3)
EDIT：图片好像显示不出来，换一张
![image](https://github.com/Hex-Dragon/PCL2/assets/69286920/7a2833c6-d86c-4854-883e-8580e26f6297)

**Reply by mc-124** [reply]
3294的方法只能识别高版本

**Reply by mc-124** [reply]
1.12.2那些找不到那玩意

**Reply by Zhichii** [reply]
那好歹也可以识别一部分愚人节快照嘛，至少不是全军覆没（

**Reply by mc-124** [reply]
但是我的方法可以识别所有版本

**Reply by Zhichii** [reply]
那我思考一下

**Reply by allMagicNB** [reply]
低版本没这文件¯\\(ツ)/¯

**Reply by Zhichii** [reply]
Recaf，启动！
话说使用net.minecraft.serverMinecraftServer.G()合法吗（法：指EULA）
![image](https://github.com/Hex-Dragon/PCL2/assets/69286920/4cd782b4-a4a5-4e9e-a18e-4dddc42fe679)

**Comment by burningtnt** [comment]
建议直接扫 Constant Pool

**Comment by Zhichii** [comment]
我觉得可以想办法反汇编net.minecraft.server.MinecraftServer
找到public String G()函数然后获取双引号中的内容
就是不知道合法不合法（Mojang EULA）

**Reply by mc-124** [reply]
一个5MB不够的VB程序塞一个反汇编？那不就成Plain Craft IDA 2 了

**Reply by Zhichii** [reply]
java反汇编比c++方便啊（而且我说了想办法）

**Reply by mc-124** [reply]
~~那不是反编译吗 怎么成反汇编了~~
你就算塞个jd也不行啊 PCL2直接变成6mb
~~一点也不轻量级~~

**Reply by mc-124** [reply]
超出启动器范畴了 谁家启动器带个反编译的

**Reply by Zhichii** [reply]
~java不是字节码（中间码）嘛，所以我说反汇编~
要是不反汇编就不反汇编了，直接找到对应位置然后读取（

**Reply by Zhichii** [reply]
**注：我没有深究过.class文件**
我发现：先打开MinecraftServer.class，有些版本没有（我还没测试）
```
#这是Python
x=open("MinecraftServer.class","rb")
y=x.read()
x.close()
```
1.5.2中：
`n=[print(_[:_.find(b"\x01")]) for _ in y.split(b"\x01\x00\x06")]`
肉眼能找到1.5.2
1.7.10中：
`n=[print(_[:_.find(b"\x01")]) for _ in y.split(b"\x01\x00\x05")]`
肉眼能找到1.7.10
1.12.2中：
`n=[print(_[:_.find(b"\x08")]) for _ in y.split(b"\x01\x00\x06")]`
肉眼能找到1.12.2
我也看不出来什么规律（而且其他版本我还没测试）
Edit：上述代码正常顺序应该是
`for i in y.split(b"\x01\x00\x0N"): print(i[0:i.find(b"\x0N")])`
（这里用n来代替无规律的数）

**Reply by burningtnt** [reply]
这就是扫 Constant Pool 的手段

**Reply by Zhichii** [reply]
眼 前 一 黑
![image](https://github.com/Hex-Dragon/PCL2/assets/69286920/a801b270-48d3-4cee-a712-592b14c5df0a)

**Reply by Zhichii** [reply]
> 1.5.2中： `n=[print(_[:_.find(b"\x01")]) for _ in y.split(b"\x01\x00\x06")]` 肉眼能找到1.5.2
> 1.7.10中： `n=[print(_[:_.find(b"\x01")]) for _ in y.split(b"\x01\x00\x05")]` 肉眼能找到1.7.10
> 1.12.2中： `n=[print(_[:_.find(b"\x08")]) for _ in y.split(b"\x01\x00\x06")]` 肉眼能找到1.12.2
> 我也看不出来什么规律（而且其他版本我还没测试）
> Edit：上述代码正常顺序应该是
> for i in y.split(b"\x01\x00\x0N"): print(i[0:i.find(b"\x0N")])
> （这里用n来代替无规律的数）

我刚看了一下[java之class文件解析](https://blog.csdn.net/qq_44778026/article/details/119997779)才发现
我写错了
并且我还发现\x01\x00\x0N中N的含义是字符串长度
所以按照刚才那个[解析](https://blog.csdn.net/qq_44778026/article/details/119997779)应该就可以扫常量池了
EDIT：那么现在的问题就是怎么从常量池里提取出版本号

**Reply by wuliaodexiaoluo** [reply]
那还不如直接算 Hash 得了
反正复制文件压缩解压 Sha1 都没变
![image](https://github.com/Hex-Dragon/PCL2/assets/160083863/585b5680-1400-44be-bf6b-e5f148eaae82)

**Reply by Zhichii** [reply]
你的意思是把每一个版本都hash一遍然后制成一个字典（Python里这么叫）？

**Reply by wuliaodexiaoluo** [reply]
每个版本 Json 不都有记录 Sha 1
历遍一下

**Reply by hejiehao** [reply]
~~直接 List 就可以了还需要 Dict？~~

**Reply by burningtnt** [reply]
> EDIT：那么现在的问题就是怎么从常量池里提取出版本号

既然你都看 HMCL 了，那就直接看 `org.jackhuang.hmcl.game.GameVersion#minecraftVersion`
分别通过 version.json，`net.minecraft.client.Minecraft`，`net.minecraft.server.MinecraftServer` 抓常量池

**Reply by hejiehao** [reply]
> 每个版本 Json 不都有记录 Sha 1 历遍一下

《历遍》

**Reply by Zhichii** [reply]
我正在写一个 Constant Pool Scanner（就是我有一种病，什么东西都想自己实现）

**Reply by MoYuan-CN** [reply]
啊？

**Reply by mc-124** [reply]
> 每个版本 Json 不都有记录 Sha 1 历遍一下

mojang会时不时更新存档 sha1不行

**Comment by wuliaodexiaoluo** [comment]
> > 每个版本 Json 不都有记录 Sha 1 历遍一下
> 
> mojang会时不时更新存档 sha1不行

Save it

**Reply by mc-124** [reply]
最好还是用releaseTime

**Comment by Zhichii** [comment]
好，现在我写出了一个我觉得还行的版本获取（我不会VB，所以以下全是Python）
```
import sys
file=open(sys.argv[1],"rb")
y=file.read()
file.close()
# 以上就是读取文件并存入y变量里，请根据需要修改
cur=8
const_cnt=y[cur]*256+y[cur+1]-1
cur+=2
const_pool=[]
for i in range(const_cnt):
    if (y[cur]==1):
        t="Utf8"
        l=y[cur+1]*256+y[cur+2]
        b=[]
        for i in range(l):
            b+=chr(y[cur+3+i])
        b="".join(b)
        const_pool.append([t,l,b])
        cur+=3+l
        continue
    if (y[cur]==3):
        cur+=5
        continue
    if (y[cur]==4):
        cur+=5
        continue
    if (y[cur]==5):
        cur+=9
        continue
    if (y[cur]==6):
        cur+=9
        continue
    if (y[cur]==7):
        cur+=3
        continue
    if (y[cur]==8):
        cur+=3
        continue
    if (y[cur]==9):
        cur+=5
        continue
    if (y[cur]==10):
        cur+=5
        continue
    if (y[cur]==11):
        cur+=5
        continue
    if (y[cur]==12):
        cur+=5
        continue
for i in const_pool:
    if('1'in i[2] or '2'in i[2] or '3'in i[2] or'4'in i[2] or '5'in i[2] or '6'in i[2] or '7'in i[2] or '8'in i[2] or '9'in i[2] or '0'in i[2]):
        print(i[2])
```
打印出来的应该就是版本号
因为我发现所有版本号都是带数字的，所以含数字的大多都是版本号
但是1.5.2有例外，所以我觉得其它版本都有例外

**Reply by Zhichii** [reply]
1.5.2会打印出两个，第一个才是版本号
EDIT：呃1.7.10例外怎么这么多
EDIT：事实证明随着版本号增大例外越来越多了

**Reply by mc-124** [reply]
有点过于复杂。。

**Reply by Zhichii** [reply]
我突然发现这一段代码有bug，在18w47**a**用不了

**Reply by Zhichii** [reply]
What？MethodHandle是什么？？？
![image](https://github.com/Hex-Dragon/PCL2/assets/69286920/491295f7-cf83-463f-890c-c4b64a36a78c)
看来那篇[解析](https://blog.csdn.net/qq_44778026/article/details/119997779)有点过时了

**Reply by burningtnt** [reply]
MethodHandle 是 Java 常量池的一部分，仅此而已。直接去抄 Constant Pool Scanner 就好了

**Reply by Zhichii** [reply]
行，这回我修了个bug，效果大概是这样的
```
Microsoft Windows [Version 10.0.19045.4170]
(c) Microsoft Corporation. All rights reserved.

H:\Minecraft> python version_getter.py MinecraftServer.class
net/minecraft/server/MinecraftServer$2
net/minecraft/server/MinecraftServer$1
java/util/Base64$Encoder
java/util/Base64
Lorg/apache/logging/log4j/Logger;
org/apache/logging/log4j/Logger
UTF_8
18w47a
Must be 64 pixels wide
org/apache/commons/lang3/Validate
Must be 64 pixels high
()Ljava/util/Base64$Encoder;
data:image/png;base64,
(Lorg/apache/logging/log4j/Logger;)V
()Lorg/apache/logging/log4j/Logger;
org/apache/logging/log4j/LogManager

H:\Minecraft>
```

**Reply by Zhichii** [reply]
稍微又改了改，有谁试试吗
```
import sys
fp = open(sys.argv[1],"rb")
clazz = fp.read()
fp.close()
# 注：clazz请自行获取，这里是一个示例
const_cnt = clazz[8] * 256 + clazz[9] - 1
cursor = 10
const_pool = []
struct_size = {0:1, 3:5, 4:5, 5:9, 6:9, 7:3, 8:3, 9:5, 10:5, 11:5, 12:5, 15:4, 16:3, 18:5, 19:3, 20:3}
for i in range(const_cnt):
    if (clazz[cursor] == 1):
        length = clazz[cursor+1]*256 + clazz[cursor+2]
        content = []
        for j in range(length):
            content.append(chr(clazz[cursor+3+j]))
        const_pool.append("".join(content))
        cursor += 3+length
    else:
        cursor += struct_size[clazz[cursor]]
v = ""
for i in const_pool:
    if('$'in i or ' 'in i or '/'in i or "UTF_8"==i): continue # 过滤
    if('1'in i or '2'in i or '3'in i or'4'in i or '5'in i or '6'in i or '7'in i or '8'in i or '9'in i or '0'in i):
        v = i
# 获取到版本号了
print(v)
```

**Reply by mc-124** [reply]
clazz？你是说class吧

**Reply by Zhichii** [reply]
class不是关键字嘛，所以用clazz代替

**Reply by Zhichii** [reply]
我研究研究VB
EDIT：应该是Visual Basic .NET

**Reply by mc-124** [reply]
> 稍微又改了改，有谁试试吗
> 
> ```
> import sys
> fp = open(sys.argv[1],"rb")
> clazz = fp.read()
> fp.close()
> # 注：clazz请自行获取，这里是一个示例
> const_cnt = clazz[8] * 256 + clazz[9] - 1
> cursor = 10
> const_pool = []
> struct_size = {0:1, 3:5, 4:5, 5:9, 6:9, 7:3, 8:3, 9:5, 10:5, 11:5, 12:5, 15:4, 16:3, 18:5, 19:3, 20:3}
> for i in range(const_cnt):
>     if (clazz[cursor] == 1):
>         length = clazz[cursor+1]*256 + clazz[cursor+2]
>         content = []
>         for j in range(length):
>             content.append(chr(clazz[cursor+3+j]))
>         const_pool.append("".join(content))
>         cursor += 3+length
>     else:
>         cursor += struct_size[clazz[cursor]]
> v = ""
> for i in const_pool:
>     if('$'in i or ' 'in i or '/'in i or "UTF_8"==i): continue # 过滤
>     if('1'in i or '2'in i or '3'in i or'4'in i or '5'in i or '6'in i or '7'in i or '8'in i or '9'in i or '0'in i):
>         v = i
> # 获取到版本号了
> print(v)
> ```

```
发生异常: KeyError
120
  File "C:\Users\124\Desktop\zc-gv.py", line 19, in <module>
    cursor += struct_size[clazz[cursor]]
KeyError: 120
```
@Zhichii 你这也不行啊

**Reply by Zhichii** [reply]
等一下，我突然想起这个项目，请问你使用的是哪个.class文件？我想看看结构

**Reply by Zhichii** [reply]
@mc-124

**Comment by mc-124** [comment]
我也用python写了个更~~简单~~的
```python
from zipfile import ZipFile
import json
from typing import List

def get_minecraft_version_id(jar_path:str):
    '''
    根据minecraft的client.jar获取minecraft的版本
    jar_path: client.jar路径
    '''
    with ZipFile(jar_path,'r') as zf:
        version_json_file = False
        minecraft_class_file = False
        server_class_file = True
        for i in zf.filelist:
            if i.filename == "version.json":
                version_json_file = i
            elif i.filename == "net/minecraft/client/Minecraft.class":
                minecraft_class_file = i
            elif i.filename == "net/minecraft/server/MinecraftServer.class":
                server_class_file = i
        else:
            if not version_json_file and not minecraft_class_file and not server_class_file:
                raise RuntimeError("无法获取传入的jar版本号")
        if version_json_file: # 新版(>=18w47b)
            zi = version_json_file
            print(1)
            with zf.open(zi,'r') as f:
                version_json = json.load(f)
                return version_json['id']
        elif minecraft_class_file: # 老版(<=1.5.2)
            zi = minecraft_class_file
            print(2)
            with zf.open(zi,'r') as f:
                data = f.read()
                #if b'Minecraft Minecraft ' not in data:
                #    continue
                d = data.split(b'Minecraft Minecraft ') # 除1.0.0rc2-2外没有任何有存档的快照有Minecraft.class
                chars = ''
                for b in d[1]:
                    if b == 1: # \x01 java常量字符串开头是\x01\x00\x{长度}
                        break
                    else:
                        chars += chr(b)
                else: # 没break
                    raise ValueError("未找到下一段个字符串的开头")
                return chars
        elif server_class_file: # 旧版(>=1.6.4)
            zi = server_class_file
            print(3)
            with zf.open(zi,'r') as f:
                data = f.read()
                if b"\x01\x00\x01.\x01\x00\x01/\x01\x00" in data: # 在1.8及更高无效
                    print(1)
                    d = data.split(b"\x01\x00\x01.\x01\x00\x01/\x01\x00")
                    chars = ''
                    for b in d[1][1:]: # 跳过第一个长度字节
                        if b == 1: # \x01
                            break
                        else:
                            chars += chr(b)
                    else:
                        raise ValueError("未找到下一段个字符串的开头")
                    return chars if chars != "RC2" else "1.0.0-rc2-2"
                else: # 这玩意我测试了接近1200遍……
                    d = data.split(b"\n") 
                    results:List[str] = []
                    for b in d:
                        if len(b) < 50 and (b'.' in b or b'w' in b or b"pre" in b or b"Pre" in b or b"Combat" in b) and (
                                b'0' in b or b'1' in b or b'2' in b or b'3' in b or b'4' in b or
                                b'5' in b or b'6' in b or b'7' in b or b'8' in b or b'9' in b
                                ):
                            c = [""]
                            enable = '0123456789qwertyuiopasdfghjklzxcvbnm.-_ ~QWERTYUIOPASDFGHJKLZXCVBNM'
                            for r in b:
                                if chr(r) in enable: # 模糊搜索
                                    c[-1]+=(chr(r))
                                else:
                                    if c[-1]:
                                        c.append("")
                            if c != [""]:
                                results += c
                    for r in results:
                        for n in "0123456789":
                            if (n in r) and ( # 过滤结果
                                'w' in r or '.' in r or '' or 'pre1' in r
                                ) and not r.startswith('.') and len(r) > 2:
                                return r
```
测试了1.20.4 1.12.2 1.12-pre7 1.10.2 1.10.2-pre1 1.8.9 1.7.10 1.6.4 1.5.2 1.4.7 1.3.2 1.2.5 18w10c 16w42a 16w02a 14w30b 13w16a 1.0-rc2-2 1.0
这个函数理论上能获取所有版本（除部分远古版外）的版本号
*注：已有vb版本 只是被折叠了*

**Reply by mc-124** [reply]
这个如果实际使用，可能需要龙猫手动翻译（
给GPT4翻译它直接不干了，我再跟它谈谈

**Reply by mc-124** [reply]
勉强转换成了不知道能不能用的~~史~~（
```vb
Imports System.IO.Compression
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq

Module MyModule
    Public Function get_minecraft_version_id(jar_path As String) As Object
        Using zf As ZipArchive = ZipFile.OpenRead(jar_path)
            Dim version_json_file As ZipArchiveEntry = Nothing
            Dim minecraft_class_file As ZipArchiveEntry = Nothing
            Dim server_class_file As ZipArchiveEntry = Nothing
            For Each i In zf.Entries
                If i.Name = "version.json" Then
                    version_json_file = i
                ElseIf i.Name = "Minecraft.class" And i.FullName.StartsWith("net/minecraft/client/") Then
                    minecraft_class_file = i
                ElseIf i.Name = "MinecraftServer.class" And i.FullName.StartsWith("net/minecraft/server/") Then
                    server_class_file = i
                End If
            Next
            If version_json_file Is Nothing And minecraft_class_file Is Nothing And server_class_file Is Nothing Then
                Throw New Exception("无法获取传入的jar版本号")
            End If
            If Not version_json_file Is Nothing Then
                Dim version_json As JObject
                Using sr As StreamReader = New StreamReader(version_json_file.Open())
                    version_json = JObject.Parse(sr.ReadToEnd())
                    Return version_json("id").ToString()
                End Using
            ElseIf Not minecraft_class_file Is Nothing Then
                Dim data() As Byte
                Using stream As Stream = minecraft_class_file.Open()
                    Dim memstream As MemoryStream = New MemoryStream()
                    stream.CopyTo(memstream)
                    data = memstream.ToArray()
                End Using
                Dim d() As String = System.Text.Encoding.UTF8.GetString(data).Split(New String() {"Minecraft Minecraft "}, StringSplitOptions.None)
                Dim chars As String = ""
                For Each b In d(1)
                    If Asc(b) = 1 Then
                        Exit For
                    Else
                        chars &= b
                    End If
                Next
                If chars = "" Then
                    Throw New Exception("未找到下一段个字符串的开头")
                End If
                Return chars
            ElseIf Not server_class_file Is Nothing Then
                Dim data() As Byte
                Using stream As Stream = server_class_file.Open()
                    Dim memstream As MemoryStream = New MemoryStream()
                    stream.CopyTo(memstream)
                    data = memstream.ToArray()
                End Using
                Dim d() As String = System.Text.Encoding.UTF8.GetString(data).Split(New String() {"\x01\x00\x01.\x01\x00\x01/\x01\x00"}, StringSplitOptions.None)
                Dim chars As String = ""
                For Each b In d(1).Skip(1)
                    If Asc(b) = 1 Then
                        Exit For
                    Else
                        chars &= b
                    End If
                Next
                If chars = "" Then
                    Throw New Exception("未找到下一段个字符串的开头")
                End If
                If chars = "RC2" Then
                    Return "1.0.0-rc2-2"
                End If
                Return chars
            End If
        End Using
        Return Nothing
    End Function
End Module
```

**Reply by Zhichii** [reply]
我试试看（
EDIT：（Python的）

**Reply by mc-124** [reply]
> 我试试看（ EDIT：（Python的）

咋样了 这么久没动静的

**Reply by Zhichii** [reply]
有可能下一个不是字符串常量吗？
![image](https://github.com/Hex-Dragon/PCL2/assets/69286920/13970652-cd2b-4be8-b00d-1cb541245317)

**Reply by mc-124** [reply]
应该不可能 看了很多class

**Reply by mc-124** [reply]
别用IDLE了 试试VS Code

**Reply by Zhichii** [reply]
话说1.12.2不是但是这时就不属于老版了
<img width="960" alt="image" src="https://github.com/Hex-Dragon/PCL2/assets/69286920/844c7b44-2543-4e1e-af44-8575f7751e8b">

**Reply by mc-124** [reply]
没听懂 你想表达什么

**Reply by Zhichii** [reply]
我的意思是我想再找找例外

**Reply by mc-124** [reply]
基本所有版本都测了个遍，没例外可找了。再说了我这个是靠换行符找的和退格号没关系
Edit: 我也没给1.12标老版啊 在这个函数里1.12属于`旧版`

**Reply by Zhichii** [reply]
话说为什么你要测模糊搜索一千多次（

**Reply by mc-124** [reply]
> 话说为什么你要测模糊搜索一千多次（

这class***的没规律可找啊
还有这模糊搜索也没太影响性能，速度也几乎感觉不到卡顿

**Reply by wuliaodexiaoluo** [reply]
你俩算力烧完没？

**Reply by wuliaodexiaoluo** [reply]
![image](https://github.com/Hex-Dragon/PCL2/assets/160083863/31eca5c1-2f9b-4461-8d9c-7b03208deb5d)
这是给你们整出来的

**Reply by Zhichii** [reply]
> > 话说为什么你要测模糊搜索一千多次（
> 
> 这class***的没规律可找啊 还有这模糊搜索也没太影响性能，速度也几乎感觉不到卡顿

不能扫常量池吗（

**Reply by Zhichii** [reply]
> ![image](https://private-user-images.githubusercontent.com/160083863/320203778-31eca5c1-2f9b-4461-8d9c-7b03208deb5d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI0MDYzNTIsIm5iZiI6MTcxMjQwNjA1MiwicGF0aCI6Ii8xNjAwODM4NjMvMzIwMjAzNzc4LTMxZWNhNWMxLTJmOWItNDQ2MS04ZDljLTdiMDMyMDhkZWI1ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwNlQxMjIwNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YTIxY2Y4YzFhMWUwM2Q5NDhiNWQ3ZmJlNTFlYmI0NzEzNjc2N2EzMmFkZTlhZDA5M2I0ODYwYWNlYjBjMzJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.pd8yGvQRRhQgKg0NB-qZDm7D4urG0ld7nKbj1qXRd5s) 这是给你们整出来的

？

**Reply by mc-124** [reply]
（

**Reply by mc-124** [reply]
我正在用C#重写我发的那段py，等写完了让猫看看能不能套进去PCL2

**Reply by allMagicNB** [reply]
> 我正在用C#重写我发的那段py，等写完了让猫看看能不能套进去PCL2

PCL2 是用 Visual Basic .Net 写的（
不过很相似罢了

**Reply by mc-124** [reply]
> > 我正在用C#重写我发的那段py，等写完了让猫看看能不能套进去PCL2
> 
> PCL2 是用 Visual Basic .Net 写的（ 不过很相似罢了

但是vb能调用C#的dll啊

**Reply by Zhichii** [reply]
EDIT：6 对不起 说话忘了小心了……

**Comment by mc-124** [comment]
@LTCatt 你看看用这玩意获取版本号行不行
https://github.com/mc-124/GetMCVersion/

**Reply by MoYuan-CN** [reply]
Visual Basic **IS NOT** C#（（（

**Reply by Zhichii** [reply]
@JingHai-Lingyun 

他说

> 但是vb能调用C#的dll啊

**Reply by MoYuan-CN** [reply]
那你等我折腾好旁加载 dll 做插件的那个 feature 的 pr 再考虑考虑吧
